### PR TITLE
Spread tests for tutorial

### PR DESCRIPTION
**Pull request by javierdelapuente**
### Overview

This PR creates a testable tutorial, in a similar fashion as it is done in `rockcraft` and `charmcraft`, using `spread` and `sphinx`. For that, it uses two plugins in `sphinx`, one to parse Markdown and one to build into Markdown, so it should work with discourse.

The tutorial had to be adapted to use env variables instead of just place holders. The scripts used were copied from the project charmcraft.

As `src-docs` is not used in NetBox (as the code is really in `paas-charm`), the `src-docs` directory was reused for it.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Review by jdkandersson — COMMENTED** [review]
I like the use of markdown instead of rst and it seems that we can include the code anyway so perhaps we use markdown at least in our docs?

**Comment by javierdelapuente** [comment]
> I like the use of markdown instead of rst and it seems that we can include the code anyway so perhaps we use markdown at least in our docs?

I think if everything is displayed correctly in discourse (pending to see), we can use this for all our charms.

File: charm/docs/tutorial/getting-started.md [hunk 1]
```diff
@@ -1,31 +1,34 @@
+<!-- This tutorial should be updated in the src-docs directory. -->
+
 # Getting Started
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:3]: I would rename the tutorial to something like 
` # Deploy the NetBox charm for the first time`

"Getting started" implies more of a how-to guide than a tutorial, and also doesn't provide details about what the tutorial experience is going to be.

File: charm/docs/tutorial/getting-started.md [hunk 2]
```diff
@@ -1,31 +1,34 @@
...
 - Create a super user.
 
-Through the process, you'll verify the workload state, and log in to
+Through the process, you’ll verify the workload state, and log in to
 your NetBox instance.
 
 ## Requirements
+
 - Juju 3 installed.
 - Juju controller that can create a model of type kubernetes.
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:20]: I would change these lines to something like 

` - Juju 3 installed and bootstrapped to a MicroK8s controller. You can accomplish this process by using a Multipass VM as outlined in this guide: [Set up / Tear down your test environment](https://juju.is/docs/juju/set-up--tear-down-your-test-environment)`

Perhaps we don't need to specify MicroK8s. It doesn't seem like the user will need to visit a website, so hopefully the rest of the tutorial remains the same if the user decides to use Multipass.

File: charm/docs/tutorial/getting-started.md [hunk 3]
```diff
@@ -1,31 +1,34 @@
...
 - Read/write access to a S3 compatible server with a bucket created.
 - Configuration compatible with the traefik-k8s charms. In the case of MicroK8S this can be achieved with the metallb addon.
 
 For more information about how to install Juju, see [Get started with Juju](https://juju.is/docs/olm/get-started-with-juju).
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:24]: If the requirements items about Juju and bootstrapping are updated, then I think this line can be removed.

File: charm/docs/tutorial/getting-started.md [hunk 4]
```diff
@@ -1,31 +1,34 @@
...
 
-
 ## Setting up a Tutorial Model
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:26]: Nitpick: Setting --> Set

File: charm/docs/tutorial/getting-started.md [hunk 5]
```diff
@@ -1,31 +1,34 @@
...
 
-To manage resources effectively and to separate this tutorial's workload from
+To manage resources effectively and to separate this tutorial’s workload from
 your usual work, we recommend creating a new model using the following command.
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:29]: Nitpick: "we recommend creating" --> "create"

File: charm/docs/tutorial/getting-started.md [hunk 6]
```diff
@@ -34,7 +37,8 @@ juju add-model netbox-tutorial
 Deploy the NetBox charm, with all its mandatory requirements (PostgreSQL, Redis and S3).
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:37]: Move this line under the `Deploy the charms` section

File: charm/docs/tutorial/getting-started.md [hunk 7]
```diff
@@ -34,7 +37,8 @@ juju add-model netbox-tutorial
...
 
 ### Deploy the charms:
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:39]: I would update this heading to say
`## Deploy the charms`

File: charm/docs/tutorial/getting-started.md [hunk 8]
```diff
@@ -43,26 +47,32 @@ storage, Redis or PostgreSQL.
 
 Set the allowed hosts. In this example every host is allowed. For a production environment
 only the used hosts should be allowed.
-```
+
+### Deploy the charms:
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:51]: Perhaps this heading should be updated to something like 
` ### Configure the NetBox charm`

Does this step need to happen before the other charms are deployed?

File: charm/docs/tutorial/getting-started.md [hunk 9]
```diff
@@ -43,26 +47,32 @@ storage, Redis or PostgreSQL.
...
+
+```bash
 juju config netbox django-allowed-hosts='*'
 ```
 
 ### Redis
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:57]: I would update this header to say 
` ### Deploy Redis`

File: charm/docs/tutorial/getting-started.md [hunk 10]
```diff
@@ -43,26 +47,32 @@ storage, Redis or PostgreSQL.
...
 
 NetBox requires Redis to work. You can deploy Redis with redis-k8s:
-```
+
+```bash
 juju deploy redis-k8s --channel=latest/edge
 ```
 
 Integrate redis-k8s with NetBox with:
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:65]: Nitpick: Remove "with"

File: charm/docs/tutorial/getting-started.md [hunk 11]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
 the local filesystem.
 
 You can configure it with:
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:87]: I would separate these commands into separate steps since a lot is going on and the `juju config` command requires the user to set some environment variables. In fact, maybe there should also be a command block and instructions for the user to set the env vars. I have included a starting point below, although maybe you will have a better explanation of the AWS variables :)

---

Deploy the s3-integrator charm:

```bash
juju deploy s3-integrator --channel edge
```

Configure the s3-integrator charm with your AWS information, and then wait for the charm to update:
```bash
juju config s3-integrator endpoint="${AWS_ENDPOINT_URL}" bucket="${AWS_BUCKET}" path=/ region="${AWS_REGION}" s3-uri-style="${AWS_URI_STYLE}"
juju wait-for application s3-integrator --query='name=="s3-integrator" && (status=="active" || status=="blocked")'
```

Once the s3-integrator charm has been deployed, integrate the charm with NetBox:

```bash
juju integrate s3-integrator netbox
```

File: charm/docs/tutorial/getting-started.md [hunk 12]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
...
 juju deploy s3-integrator --channel edge
-juju config s3-integrator endpoint="<aws_endpoing_url>" bucket=<bucket_name> path=<path_in_bucket> region=<region> s3-uri-style=<path_or_host>
+juju config s3-integrator endpoint="${AWS_ENDPOINT_URL}" bucket="${AWS_BUCKET}" path=/ region="${AWS_REGION}" s3-uri-style="${AWS_URI_STYLE}"
 juju wait-for application s3-integrator --query='name=="s3-integrator" && (status=="active" || status=="blocked")'
-juju run s3-integrator/leader sync-s3-credentials access-key=<aws_access_key_id> secret-key=<aws_secret_access_key>
+juju run s3-integrator/leader sync-s3-credentials access-key="${AWS_ACCESS_KEY_ID}" secret-key="${AWS_SECRET_ACCESS_KEY}"
 juju integrate s3-integrator netbox
 ```
 
 See the [s3-integrator charmhub page](https://charmhub.io/s3-integrator) for more information.
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:97]: Nitpick: charmhub --> Charmhub

File: charm/docs/tutorial/getting-started.md [hunk 13]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
...
 
 You can check the proxied endpoints with the command:
-```
+
+```bash
 juju run traefik-k8s/0 show-proxied-endpoints --format=yaml
+# [docs:traefik-show-endpoints-end]
 ```
 
 ## Create superuser
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:122]: I would rephrase this header to say
`## Create a super user`

File: charm/docs/tutorial/getting-started.md [hunk 14]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
...
+
 To be able to login to NetBox, you can create a super user with the next command:
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:124]: login --> log in

File: charm/docs/tutorial/getting-started.md [hunk 15]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
...
-```
-juju run netbox/0 create-superuser username=<admin_username> email=<admin_email>
+
+```bash
+juju run netbox/0 create-superuser username=admin email=admin@example.com
 ```
 
-Congratulations, With the username created and the password provided in the response, 
+Congratulations, With the username created and the password provided in the response,
 you have now full access to your own NetBox!
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:131]: Two comments here:
1. Does it make sense to instruct the user to visit http://netbox_hostname/netbox-tutorial-netbox and check that they have full access?
2. After this section, I recommend adding a `## Clean up the environment` section where the user destroys the Juju model and deletes the Multipass VM (if they used Multipass)

File: charm/docs/tutorial/getting-started.md [hunk 16]
```diff
@@ -75,41 +85,47 @@ This allows uploaded images to be placed on an S3 compatible server instead of
...
 juju deploy traefik-k8s --channel edge --trust
-juju config traefik-k8s external_hostname=<netbox_hostname>
-juju config traefik-k8s routing_mode=path
+juju config traefik-k8s external_hostname=${EXTERNAL_HOSTNAME}
+juju config traefik-k8s routing_mode=${ROUTING_MODE}
 juju integrate traefik-k8s netbox
 ```
 
 If the host `netbox_hostname` can be resolved to the correct IP (the load balancer IP),
 you should be able to browse NetBox in the url http://netbox_hostname/netbox-tutorial-netbox
```
Comment (erinecon) [inline charm/docs/tutorial/getting-started.md:113]: url --> URL

File: charm/docs/tutorial/getting-started.md (hunk 3 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:24]: That line was automatically added by sphinx.

File: charm/docs/tutorial/getting-started.md (hunk 6 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:37]: I have changed it a bit, as it was badly ordered..

File: charm/docs/tutorial/getting-started.md (hunk 7 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:39]: I have changed it a bit, as it was badly ordered..

File: charm/docs/tutorial/getting-started.md (hunk 8 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:51]: I have put a new NetBox header that includes both things

File: charm/docs/tutorial/getting-started.md (hunk 9 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:57]: I have updated all the titles of that section

File: charm/docs/tutorial/getting-started.md (hunk 15 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:140]: 1. What would you suggest instead?

Comment (erinecon) [reply charm/docs/tutorial/getting-started.md:140]: After the "Congratulations" sentence, I would add one more sentence instructing the user to visit http://netbox_hostname/netbox-tutorial-netbox and log in.

File: charm/docs/tutorial/getting-started.md (hunk 3 above)
Comment (erinecon) [reply charm/docs/tutorial/getting-started.md:24]: Ah okay, we can leave it alone then.

**Review by erinecon — COMMENTED** [review]
Thanks for updating the tutorial and adding a spread test! The documentation changes look good to me :+1: 
(Not giving an official approval since code was also changed in this PR.)

**Comment by alithethird** [comment]
@javierdelapuente Is this still in progress?

**Comment by javierdelapuente** [comment]
It is, but can be closed and reopen when appropriate. Thanks!
